### ISSUE TITLE
[ML/Dynamics] General-purpose substrate for learned, probabilistic, continuous-time computation

### ISSUE THREAD

**Issue by dancxjo** [dancxjo/conduit#2132]
## Outcome

Extend Conduit so it can faithfully describe, execute, observe, and reproduce modern learned dynamical systems without turning the Conduit kernel or semantic catalog into a second PyTorch/JAX.

The motivating proof is **Tongues**: learn a shared latent dynamical account of speech from paired and unpaired acoustic/articulatory data, support inference in either direction, preserve uncertainty, and allow later scientific analysis of learned trajectories and couplings.

This is an architecture roadmap, not one implementation PR.

## Background

The Tongues direction has converged on a stronger model of speech than a sequence of phoneme-sized boxes. Physical speech is produced by multiple partially independent but coordinated trajectories: tongue, lips, jaw, velum, larynx, F0, etc. Their gestures overlap in time, interact, and need not share identical boundaries. A learned system should therefore be able to infer continuous latent dynamics from training data rather than having segmentation imposed as the substrate.

The desired ML architecture is approximately:

```text
                       speaker/anatomy/style
                              |
audio -> audio encoder -------|
                              v
                         latent z(t)
                              ^
articulation -> encoder -------|
                              |
               +--------------+--------------+
               |                             |
        acoustic decoder              articulatory decoder
               |                             |
             audio                        trajectories
```

Training should permit:

- audio -> latent -> audio
- articulation -> latent -> articulation
- audio -> latent -> articulation
- articulation -> latent -> audio
- paired modalities constraining the same latent process
- missing modalities
- probabilistic posterior inference when the inverse problem is non-unique
- recurrent/stateful updates across time
- continuous-time integration where appropriate
- model training and checkpoint production without exposing framework-internal autograd graphs as Conduit semantics
- scientific post-analysis such as phase/coupling discovery, probes, and sparse dynamical-law discovery

## Existing Conduit strengths to preserve

Conduit already has much of the surrounding machinery:

- semantic Kinds and renderer/runtime-independent Gear contracts;
- explicit bounded values and resource references;
- quantities and temporal quantities;
- canonical PCM blocks with clock identity, start frame, sample rate, and discontinuity;
- stateful `Operation` execution (`resume`, `advance`, bounded host requests);
- explicit host-operation admission and resource accounting;
- model identity/lifecycle/resource concepts in existing local-model support;
- debug observation and Patchbay projection machinery;
- ordinary data dependencies intentionally remain acyclic.

The plan should **generalize these strengths**, not create an ML side-runtime.

### 1. Conduit orchestrates ML; it does not reimplement ML frameworks

Torch, JAX, Burn, Candle, ONNX Runtime, custom Rust, etc. may realize semantic operations at Host edges. Conduit should own meaning, resource admission, provenance, identity, routing, state transitions, and observable evidence.

Do not begin by adding primitive Gears for every matmul, convolution, layer norm, backward pass, or optimizer kernel.

### 2. Segmentation is a projection, not the ontology

Time-indexed signals and latent trajectories must be representable without asserting that their sampled frames are the underlying linguistic units. Phones, syllables, morae, word boundaries, and IPA annotations should be able to appear later as labels/probes/views over learned continuous structure.

### 3. Feedback must cross an explicit state/time boundary

Do not relax the planner into arbitrary cyclic dataflow. Recurrent computation should become legal through explicit state/delay semantics so the graph remains acyclic within one logical step/instant.

### 4. Bidirectional does not mean mathematically bijective

Speech acoustics -> articulation is ill-posed: multiple physical configurations may explain similar acoustics. The semantic model must support distributions/posteriors and multiple samples, not claim a unique inverse.

### 5. A model may represent a relation, not only a directed function

One learned artifact may model a joint relationship among acoustics, articulation, and latent dynamics. Direction belongs to the query (`condition`, `infer`, `sample`, `reconstruct`), not necessarily to separate artifact identities such as `AudioToArticulationModel` and `ArticulationToAudioModel`.

### 6. Scientific provenance matters

Training corpus identity, synchronization, coordinate systems, calibration, masking, splits, model/checkpoint identity, random seed, runtime realization, resource envelope, and metrics must remain distinguishable and inspectable.

## Planned child slices

This roadmap should be implemented as reviewable child issues covering at least:

1. explicit `State`/`Delay` semantics and planner-safe recurrent feedback;
2. tensor/n-dimensional numeric value semantics with named/typed axes;
3. generic sampled signal / trajectory semantics generalized from PCM;
4. provider-neutral model artifact, signature, state, and checkpoint semantics;
5. probabilistic outputs, distributions, samples, uncertainty, and seed semantics;
6. multimodal observation/corpus synchronization, calibration, coordinate frames, and missing data;
7. training/evaluation sessions with Host-owned autodiff/optimizer internals and durable checkpoint outputs;
8. continuous-time dynamics/integration contracts (ODE/SDE-style realization at Host edges);
9. relational/bidirectional model query semantics;
10. Host execution adapters and accelerator/resource admission for ML workloads;
11. Patchbay real-time learned-system observation/debugging projections;
12. Tongues capstone experiment that trains without phone segmentation and tests whether useful gestural/dynamical structure emerges;
13. scientific analysis/probing of learned latent trajectories, including phase/coupling analysis and sparse/interpretable dynamics discovery.

## Explicit non-goals

- no arbitrary cyclic ordinary dataflow;
- no mandatory CUDA/PyTorch dependency in architecture crates;
- no rebuilding autograd in the Conduit kernel;
- no assumption that tensors are inherently ML-specific;
- no assumption that every signal is uniformly sampled forever;
- no assertion that a posterior sample is ground-truth physical articulation;
- no conflation of model artifact identity, model parameters, optimizer state, training run, dataset, or checkpoint;
- no Tongues-only concepts in generic architecture when a broader numeric/temporal/probabilistic concept suffices;
- no phoneme/IPA segmentation requirement in the foundational representation.

## Acceptance for the roadmap

This umbrella is complete when the child slices collectively make it possible to express and prove an end-to-end Form that:

1. consumes synchronized audio and articulatory observations;
2. represents them as time-indexed numeric signals with explicit clocks/coordinate semantics;
3. runs a model through a provider-neutral Host capability;
4. maintains recurrent training/model state through explicit state boundaries rather than arbitrary graph cycles;
5. returns probabilistic latent/articulatory predictions;
6. updates model state in a bounded training step and emits a durable checkpoint;
7. can query the same learned artifact in both acoustic->articulatory and articulatory->acoustic directions;
8. can integrate learned continuous dynamics where selected by the realization;
9. preserves dataset/model/runtime/resource/provenance evidence;
10. exposes useful live values, trajectories, uncertainty, losses, and state transitions in Patchbay;
11. supports a Tongues proof in which linguistic segment labels are withheld during representation learning and introduced afterward only for analysis.

## Architectural test

If this work succeeds, the following sentence should be true:

> **Conduit can host learned dynamical science because its abstractions are general enough for stateful, continuous, probabilistic, multimodal computation, not because the core learned the special case called “neural network.”**

**Comment by dancxjo** [comment]
## Created child roadmap

The architecture/research slices for this roadmap are now tracked as:

1. #2133 — explicit State/Delay semantics for planner-safe recurrence
2. #2134 — canonical tensors with semantic axes and bounded payload/reference rules
3. #2135 — generic sampled Signal/Trajectory semantics generalized from PCM timing
4. #2137 — provider-neutral ModelArtifact/signature/state/checkpoint semantics
5. #2138 — probabilistic results, samples, uncertainty, and random-seed semantics
6. #2139 — multimodal ObservationSet/corpus semantics: clocks, calibration, coordinate frames, masks, splits, provenance
7. #2140 — bounded Training/Evaluation sessions with Host-owned autodiff/optimizer internals
8. #2141 — continuous-time dynamics/integration contracts
9. #2142 — queryable learned relations for conditional/bidirectional use without bidirectional Cords
10. #2143 — Host ML execution adapters plus accelerator/resource admission
11. #2144 — Patchbay watches for tensors, trajectories, uncertainty, state, loss, and dynamics
12. #2145 — Tongues capstone: learn shared latent speech dynamics from audio + articulation without phone segmentation
13. #2146 — Tongues scientific analysis: coupling, emergent units, probes, and interpretable dynamical-law discovery

### Recommended dependency spine

```text
#2133 State/Delay
#2134 Tensor
   |
   +--> #2135 Signal/Trajectory --> #2139 ObservationSet/corpus
   |
   +--> #2137 ModelArtifact --> #2138 Probability
                    |              |
                    +------> #2142 relational queries
                    |
#2133 + #2139 + #2138 + #2137 --> #2140 training/evaluation
#2133 + #2135 + #2137 ----------> #2141 continuous dynamics
#2137 + #2140 + #2141 + #2142 --> #2143 Host execution/adapters

all observable semantic slices --------> #2144 Patchbay watches
substrate + Host + watch slices --------> #2145 Tongues capstone
#2145 + trajectory/dynamics/watch ------> #2146 scientific discovery/analysis
```

The intended architectural stop line throughout is: **Conduit owns semantic meaning, state/time boundaries, resource admission, identity, provenance, routing, and observation; selected Host runtimes own autograd, tensor allocators, accelerator kernels, numerical solvers, and framework-specific model internals.**

**Comment by dancxjo** [comment]
Roadmap accepted and complete at exact main `5922405a108c947c038dcaa967f28c94bd9744d1`.

All thirteen owned child issues are closed: #2133, #2134, #2135, #2137, #2138, #2139, #2140, #2141, #2142, #2143, #2144, #2145, and #2146. (Number #2136 is a historical merged pull request, not a missing roadmap child.)

The collected exact-main surface now establishes the umbrella contract: explicit recurrent State/Delay boundaries; finite typed tensors and sampled trajectories; distinct provider-neutral model/checkpoint identities; probabilistic bounded results; synchronized corpus/calibration/mask/split provenance; bounded Host-owned training and model compute; continuous-dynamics and relational query contracts; real authoritative Patchbay learned Watches; and the Tongues paired-latent experiment plus post-freeze scientific analysis.

The final merge #2237 passed required exact-head `check` with affected workspace shards only (https://github.com/dancxjo/conduit/actions/runs/33710877021) and the real Patchbay browser proof 5/5 (https://github.com/dancxjo/conduit/actions/runs/33710877015). Fresh exact-main validation then passed all five Tongues tests, regenerated identity-bound analysis report `sha256:5e86b11476fa1af0ae6d51cab949dfb08ffb652c18d564ae9c3d7e80a357affd`, rebuilt the whole Patchbay HTML package after targeted cleanup, and passed the focused browser proof 5/5.

The roadmap's scientific stop line remains intact. The final analysis preserves its failed reversed-pair control, partial event alignment, single-speaker limitation, and non-identifiable C-center/front-end/cross-speaker questions; it does not turn numerical phase, clusters, or sparse predictive equations into oscillator, linguistic-unit, or causal claims. Conduit owns the finite semantic, identity, admission, state, routing, and observation contracts while numerical implementation remains at the Host edge.
